Source/WebCore/page/Navigation.cpp

```diff
@@ -383,23 +378,29 @@ bool Navigation::hasEntriesAndEventsDisabled() const
 }
 
 // https://html.spec.whatwg.org/multipage/nav-history-apis.html#resolve-the-finished-promise
-void Navigation::resolveFinishedPromise(const NavigationAPIMethodTracker& apiMethodTracker)
+void Navigation::resolveFinishedPromise(NavigationAPIMethodTracker* apiMethodTracker)
 {
-    // FIXME: We should be able to assert committedToEntry is always set.
-    if (apiMethodTracker.committedToEntry) {
-        apiMethodTracker.committedPromise->resolve<IDLInterface<NavigationHistoryEntry>>(*apiMethodTracker.committedToEntry);
-        apiMethodTracker.finishedPromise->resolve<IDLInterface<NavigationHistoryEntry>>(*apiMethodTracker.committedToEntry);
+    if (!apiMethodTracker->committedToEntry) {
+        apiMethodTracker->finishedBeforeCommit = true;
+        return;
     }
+
+    apiMethodTracker->committedPromise->resolve<IDLInterface<NavigationHistoryEntry>>(*apiMethodTracker->committedToEntry);
+    apiMethodTracker->finishedPromise->resolve<IDLInterface<NavigationHistoryEntry>>(*apiMethodTracker->committedToEntry);
     cleanupAPIMethodTracker(apiMethodTracker);
 }
 
 // https://html.spec.whatwg.org/multipage/nav-history-apis.html#notify-about-the-committed-to-entry
-static void notifyCommittedToEntry(NavigationAPIMethodTracker& apiMethodTracker, NavigationHistoryEntry* entry)
+void Navigation::notifyCommittedToEntry(NavigationAPIMethodTracker* apiMethodTracker, NavigationHistoryEntry* entry)
 {
     ASSERT(entry);
-    apiMethodTracker.committedToEntry = entry;
+    apiMethodTracker->committedToEntry = entry;
     // FIXME: 2. If apiMethodTracker's serialized state is not null, then set nhe's session history entry's navigation API state to apiMethodTracker's serialized state.
-    apiMethodTracker.committedPromise->resolve<IDLInterface<NavigationHistoryEntry>>(*entry);
+
+    if (apiMethodTracker->finishedBeforeCommit)
```

Comment:
I am curious, is the finishedBeforeCommit logic described in the specification or something we use internally to make things work?

Reply:
Unfortunately its not in the spec, but its exactly what Chromium does. And this is the only thing that makes sense, the spec as described just has this situation which would always fail.

I can look into contributing back to the spec.

Reply:
Thanks, I kept checking the spec to see if I missed something for this stop condition when I was trying to make the markAsHandled PR work.